src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala

```diff
@@ -0,0 +1,283 @@
+    if (entities.isEmpty) DBIO.successful(0)
+    else dataAccess.entityQuery.save(SlickWorkspaceContext(workspace), entities)
+  }
+
+  def attachOutputs(workspace: Workspace, workflowsWithOutputs: Seq[(WorkflowRecord, ExecutionServiceOutputs)], entitiesById: Map[UUID, Entity], outputExprepressions: Map[String, String]): Seq[Either[(Option[Entity], Option[Workspace]), (WorkflowRecord, Seq[AttributeString])]] = {
+    workflowsWithOutputs.map { case (workflowRecord, outputsResponse) =>
+      val outputs = outputsResponse.outputs
+
+      val attributes = outputExprepressions.map { case (outputName, attributeName) =>
+        Try {
+          attributeName.value -> outputs.getOrElse(outputName, {
+            throw new RawlsException(s"output named ${outputName} does not exist")
+          })
+        }
+      }
+
+      if (attributes.forall(_.isSuccess)) {
+        Left(updateEntityAndWorkspace(entitiesById(workflowRecord.workflowEntityId.get), workspace, attributes.map(_.get).toMap))
+
+      } else {
+        Right((workflowRecord, attributes.collect { case Failure(t) => AttributeString(t.getMessage) }.toSeq))
+      }
+    }
+  }
+
+  def updateEntityAndWorkspace(entity: Entity, workspace: Workspace, workflowOutputs: Map[String, Attribute]): (Option[Entity], Option[Workspace]) = {
+    //Partition outputs by whether their attributes are entity attributes (begin with "this.") or workspace ones (implicitly; begin with "workspace.")
+    //This assumption (that it's either "this." or "workspace.") will be guaranteed by checking of the method config when it's imported; see DSDEEPB-1603.
+    //Yes I know this is a var but it's more terse this way.
+    var (entityAttributes, workspaceAttributes) = workflowOutputs.partition({ case (k, v) => k.startsWith("this.") })
+    entityAttributes = entityAttributes.map({ case (k, v) => (k.stripPrefix("this."), v) })
+    workspaceAttributes = workspaceAttributes.map({ case (k, v) => (k.stripPrefix("workspace."), v) })
+
+    val updatedEntity = if (entityAttributes.isEmpty) None else Option(entity.copy(attributes = entity.attributes ++ entityAttributes))
+    val updatedWorkspace = if (workspaceAttributes.isEmpty) None else Option(workspace.copy(attributes = workspace.attributes ++ workspaceAttributes))
+    
+    (updatedEntity, updatedWorkspace)
+  }
+
+  def saveErrors(errors: Seq[(WorkflowRecord, Seq[AttributeString])], dataAccess: DataAccess) = {
+    DBIO.sequence(errors.map { case (workflowRecord, errorMessages) =>
+      dataAccess.workflowQuery.filter(_.id === workflowRecord.id).update(workflowRecord.copy(status = WorkflowStatuses.Failed.toString)) andThen
+        dataAccess.workflowQuery.saveMessages(errorMessages, workflowRecord.id)
+    })
+  }
+
+  private def getUserInfo = {
+    val expiresInSeconds: Long = Option(credential.getExpiresInSeconds).map(_.toLong).getOrElse(0)
+    if (expiresInSeconds <= 5*60) {
```

Comment:
Should this be configurable?  If so, make a JIRA to do it later rather than bothering for now.